### DISCUSSION TITLE
Example for using LiquidMenu inside classes

### DISCUSSION THREAD

**Discussion by ghost** [VasilKalchev/LiquidMenu#87] [Q&A] [answered]
The examples provided are not useful for a real-case scenario. Usually, any application is organized in classes and it's not a good approach to put everything outside.

My limited knowledge of C++ does not allow me to fix the syntax. I provide here some code (far from to be elegant, but more usable than the examples) and I kindly ask you a minimal example to show how to integrate this great library in a class.

main.cpp

```
const int rs = 3, en = 0, d4 = 4, d5 = 30, d6 = 12, d7 = 6;
LiquidCrystal lcd(rs, en, d4, d5, d6, d7);

Menu menu(&lcd);

void setup() { }
void loop() { }
```

myclass.h

```
class Menu
{
public: 
    Menu(LiquidCrystal *lcd);        

private:
    int _myvar;

    LiquidCrystal *_lcd;
    LiquidMenu _menu;
    LiquidScreen _screenSetCC1;

    LiquidLine line1_setupCC1;
    LiquidLine line2_setupCC1;

    void editIntPlus();
};
```

myclass.cpp

```
Menu::Menu(LiquidCrystal *lcd) : _lcd(lcd), _menu(*_lcd),
    line1_setupCC1(0, 0, "Set CC #1"),
    line2_setupCC1(0, 1, _myvar)
{
    
    line2_setupCC1.attach_function(1, editIntPlus);

    _screenSetCC1.add_line(line1_setupCC1);
    _screenSetCC1.add_line(line2_setupCC1);
    _menu.add_screen(_screenSetCC1);
}

void Menu::editIntPlus()
{
    _myvar++;
}
```

This code does not compile:

> error: invalid use of non-static member function 'void Menu::editIntPlus()'

But I don't want to use static members, otherwise I cannot access to the other variables or functions.
I tried this:

`line2_setupCC1.attach_function(1, (void (*)()) &this->editIntPlus);`

it compiles but when I call the function:

`Serial.println(_menu.call_function(1));`

it returns 0, so the actual function is not called.

**Please, can you show me how to deal with classes?**

Furthermore, in my menu I have dozens of vars to set. I really don't want to create a lot of functions to increase or decrease each single var! Is there a way to retrieve the variable being edited inside the callback? In this way I can set the same function for all the variables (given they are of the same type) ending in a more elegant a re-usable code.

**Comment by VasilKalchev** [comment, marked as answer]
Hello, @Mark-81.

The problem is that the library holds pointers to functions and not pointers to member functions of your specific class. That's why you must trick the compiler by casting your member function address to a normal function address. You can read more about member function pointers [here](https://opensource.com/article/21/2/ccc-method-pointers).

On your point about using classes for everything. If you're coming from Java or C#, it's not bad practice to have declarations outside of classes in C++. If you're new to coding, don't take guidelines as rules (and I'm pretty sure there isn't such a guideline). If you ask me, one of the best advices is to not write needlessly complex code.


> ...I have dozens of vars to set. I really don't want to create a lot of functions to increase or decrease each single var! Is there a way to retrieve the variable being edited inside the callback? In this way I can set the same function for all the variables...

Using a single callback function and figuring out who called it, will make that function huge, which doesn't tick the elegant box. A workaround would be to group your callback functions inside a namespace, spanning multiple files or in a separate one, just for callback functions. But if you still want to do it in a single callback function, you can use `LiquidScreen * LiquidMenu::get_currentScreen()` and `uint8_t LiquidMenu::get_focusedLine()` to figure out from which screen and which line the function was called.

Hope this helps.
